### PR TITLE
feat: pass Retry-After to end user

### PR DESCRIPTION
**Pull request by lidel**
This PR aims to use cooldown and seturn it in `Retry-After` header to the client.

Based on https://github.com/ipfs/go-libipfs/pull/194

### CODE REVIEW

File: blockstore.go [hunk 1]
```diff
@@ -68,8 +68,10 @@ func (e *exchangeBsWrapper) Close() error {
 }
 
 // gatewayError translates underlying blockstore error into one that gateway code will return as HTTP 502 or 504
+// it also makes sure Retry-After hint from remote blockstore will be passed to HTTP client, if present.
 func gatewayError(err error) error {
-	if errors.Is(err, gateway.ErrGatewayTimeout) || errors.Is(err, gateway.ErrBadGateway) {
+	if errors.Is(err, &gateway.ErrorResponse{}) ||
+		errors.Is(err, &gateway.ErrorRetryAfter{}) {
```
Comment (lidel) [inline blockstore.go:74]: :information_source: this removes the need for listing every error code

File: blockstore.go [hunk 2]
```diff
@@ -82,6 +84,16 @@ func gatewayError(err error) error {
...
 
+	// Saturn errors with RetryAfter hint need to be converted to Gateway ones
+	var saturnTooManyRequests *caboose.ErrSaturnTooManyRequests
+	if errors.As(err, &saturnTooManyRequests) {
+		return gateway.NewErrorRetryAfter(saturnTooManyRequests, saturnTooManyRequests.RetryAfter)
+	}
+	var saturnCidCoolDown *caboose.ErrCidCoolDown
+	if errors.As(err, &saturnCidCoolDown) {
+		return gateway.NewErrorRetryAfter(saturnCidCoolDown, saturnCidCoolDown.RetryAfter)
+	}
```
Comment (lidel) [inline blockstore.go:95]: :thought_balloon: @hacdias any ideas how we can avoid this mapping?  make caboose return error which is already wrapped in  `ErrorRetryAfter` ?

Comment (hacdias) [reply blockstore.go:95]: Yes, if caboose already returns `ErrorRetryAfter`, we won't need this.

Comment (hacdias) [reply blockstore.go:95]: https://github.com/filecoin-saturn/caboose/pull/60

**Review by hacdias — COMMENTED** [review]
LGTM, but let's wait for https://github.com/filecoin-saturn/caboose/pull/60.

File: blockstore.go [hunk 3]
```diff
@@ -82,6 +84,14 @@ func gatewayError(err error) error {
...
 	}
 
+	// (Saturn) errors that support the RetryAfter interface need to be converted
+	// to the correct gateway error, such that the HTTP header is set.
+	for v := err; v != nil; v = errors.Unwrap(v) {
+		if r, ok := v.(interface{ RetryAfter() time.Duration }); ok {
+			return gateway.NewErrorRetryAfter(err, r.RetryAfter())
+		}
+	}
+
```
Comment (hacdias) [inline blockstore.go:94]: Looks very hacky, but I love it. Now we support all errors that export the `RetryAfter() time.Duration` interface.

**Review by hacdias — APPROVED** [review]

**Comment by hacdias** [comment]
I added a test so I am confident that the conversion works. I will merge once CI is green.

**Pull request merged**